Implement SDR classifier in NuPIC

ywcui1990 opened the pull request:
@subutai @scottpurdy 

Fixes: https://github.com/numenta/nupic/issues/3053

- Implement SDR classifier in Nupic
- Add unit tests for SDR classifier. These tests are similar to the cla_classifier_test, except for the obsolete ones and the testWriteRead (due to lack of SDRClassifierProto)

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
```
2016

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+    return retval
+
+
+  def inferSingleStep(self, patternNZ, weightMatrix):
```
Docstring?

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+# ----------------------------------------------------------------------
+
+"""
+Implementation for a SDR classifier.
```
"Implementation of a SDR classifier"

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+  encoders (the "classification") describing the true (target) input.
+
+  The SDR classifier maps input patterns to class labels. There are as many
+  output units as the number of class labels (buckets). The output is a
```
"(or buckets in the case of scalar encoders)"

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+    @param steps (list) Sequence of the different steps of multi-step
+        predictions to learn
+    @param alpha (float) The alpha used to adapt the weight matrix during
+        learning. A lower alpha results in longer term memory.
```
Does alpha really control how long the memory is stored? (If an SDR is not seen for a really long time it will not forget it regardless of alpha.) It seems like alpha just controls how fast it adapts?

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+      if self._actualValues[bucketIdx] is None:
+        self._actualValues[bucketIdx] = actValue
+      else:
+        if isinstance(actValue, int) or isinstance(actValue, float):
```
Do we need to check for "long" as well?

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+    @param verbosity (int) verbosity level, can be 0, 1, or 2
+    """
+    # Save constructor args
+    self.steps = steps
```
We should check that steps, alpha, and actValueAlpha are legal values and expected types. (e.g. alpha should be in [0,1]).

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,354 @@
+  def inferSingleStep(self, patternNZ, weightMatrix):
+    outputActivation = numpy.zeros(self._maxBucketIdx + 1)
+    for bit in patternNZ:
+      outputActivation += weightMatrix[bit, :]
```
Can we do something like outputActivation += weightMatrix[patternNZ].sum()?  It might be faster than the loop.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,445 @@
+
+"""Unit tests for SDRClassifier module."""
+
+CL_VERBOSITY = 0
```
This is not used anywhere.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,445 @@
+    self.assertEqual(retval['actualValues'][index], value)
+
+
+  def _checkProbabilityGreater(self, retval, index, probability):
```
Not used anywhere.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,445 @@
+    self.assertGreater(retval[1][0], 0.9)
+
+
+  def testSingleValue0Steps(self):
```
Can we have a test with steps=[0] that distinguishes between steps=[1]? If we alternate the classification between two buckets, 0-step should learn one of them and 1-step should learn the other.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,445 @@
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""Unit tests for SDRClassifier module."""
```
We should have tests that ensure exceptions are raised properly when bad values are passed in.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,445 @@
+
+  def testInitialization(self):
+    c = self._classifier([1], 0.1, 0.1, 0)
+    self.assertEqual(type(c), self._classifier)
```
We can add some tests of constructors with bad parameters - they should throw exceptions.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,445 @@
+    self.assertAlmostEqual(result['actualValues'][4], 34.7, places=5)
+
+
+  def testComputeComplex(self):
```
We don't see to have test cases where learn=False and infer=True, or vice versa.  Also, if both are False, an exception should be raised.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,445 @@
+              learn=True, infer=True)
+    serialized = pickle.dumps(c)
+    c = pickle.loads(serialized)
+    result = c.compute(recordNum=4,
```
We should also check that the member variables are set as expected after deserialization.  For example, alpha could be incorrect after unpickling, and this test would not catch that.

subutai commented:
Done with my initial review. Overall the code looks very clean and well written. Left suggestions for more tests plus some more we discussed on paper.

ywcui1990 commented:
@subutai 

Thanks for the code review. I have incorporated your comments and added 3 additional tests, which are 
(1) test that the classifier captured the probability distribution given enough repetition 
(2) test that the classifier are robust to shared bits in the input SDRs and 
(3) test that the classifier adapts to changes in the data streams and remembers the unchanged part

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,711 @@
+                learn=True, infer=True)
+      recordNum += 1
+
+      randomNumber = random.random()
```
In this second phase, can we leave out the second mapping completely? The classifier should not touch the bits for SDR2 so it should still map to 50%/50%.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,711 @@
+                learn=True, infer=True)
+      recordNum += 1
+
+    for _ in xrange(30000):
```
Do we really need 10,000 and 30,000 iterations for this test? It would be good if we could do this much quicker.

subutai commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,711 @@
+        bucketIdx = 1
+      else:
+        bucketIdx = 2
+      c.compute(recordNum=recordNum, patternNZ=SDR1,
```
You can probably set infer to False here.

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,375 @@
+  def calculateError(self, classification):
+    """
+    Calculate error signal
+    :param classification:
```
Should use a consistent docstring style, i.e. @param

scottpurdy commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,375 @@
+g_debugPrefix = "SDRClassifier"
+
+
+def _pFormatArray(array_, fmt="%.2f"):
```
Put this at the end of the file. It just gets in the way up here

scottpurdy commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,375 @@
+  During learning, the connection weights between input units and output units
+  are adjusted to maximize the likelihood of the model
+
+  References:
```
Mention that it is a slight variation of the previous CLAClassifier that was not based on this reference. Also include a simple code snippet showing the usage of the class (instantiation and getting predictions out).

scottpurdy commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,375 @@
+
+import numpy
+
+g_debugPrefix = "SDRClassifier"
```
Remove this. It is only used once and you can just put the string there (or use `type(self).__name__`)

scottpurdy commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,375 @@
+    Architectures and Applications, pp 227-236, Springer-Verlag, 1990
+  """
+
+  __VERSION__ = 1
```
All caps without underscores is probably better for this (matches our naming conventions)

scottpurdy commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,375 @@
+    return predictDist
+
+
+  def calculateError(self, classification):
```
Is this part of the public interface? If it is only meant to be used internally then include a single underscore prefix: `_ calculateError`

scottpurdy commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,711 @@
+
+
+
+if __name__ == '__main__':
```
double quotes

scottpurdy commented on tests/unit/nupic/algorithms/sdr_classifier_test.py:
```diff
@@ -0,0 +1,711 @@
+
+  @staticmethod
+  def _compute(classifier, recordNum, pattern, bucket, value):
+    classification = {'bucketIdx': bucket, 'actValue': value}
```
double quotes here and elsewhere

scottpurdy commented:
done reviewing

ywcui1990 commented:
@subutai @scottpurdy Thanks for the code review for both of you. 

I incorporated your feedback. In addition, 
* Added a test for simultaneous multi-step prediction
* Sped up the unit tests. It takes less than 2 secs to run all the tests on my computer
* Allowed classification to be "None" in pure inference mode

scottpurdy commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,393 @@
+  # inference
+  result = c.compute(recordNum=1, patternNZ=[1, 5, 9],
+                     classification={"bucketIdx": 4, "actValue": 34.7},
+                     learn=False, infer=True)
```
The tricky part is interpreting `result`. Can you follow this with a loop that prints each prediction and associated predicted value (only do this for one of the steps). For example, something like this:

```
# Print the top three predictions for 5 steps out.
for probability, value in sorted(zip(result[5], result["actualValues"]), reverse=True)[:3]:
  print "Prediction of {} has probability of {}.".format(value, probability*100.0)
```

subutai commented on src/nupic/algorithms/sdr_classifier.py:
```diff
@@ -0,0 +1,393 @@
+
+  Example Usage:
+
+  c = SDRClassifier(steps=[0], alpha=0.1, actValueAlpha=0.1, verbosity=0)
```
I think it would be better to use steps=[1] in this example - that's the most common case.

subutai commented:
One minor comment and agree with @scottpurdy 's comment. I have no further issues - all looks good to me! :+1:

ywcui1990 commented:
I have fixed the example code snippet. It should be ready to merge if you have no further comments @scottpurdy

cogmission commented:
@ywcui1990 Please see https://github.com/numenta/nupic/issues/3067
